Added the ability for different boundary values for different frames

ghaarsma opened the pull request:
Small modification in the linear boundary filter to have correct padding.

Also added the ability to have different linear boundary values for different frames. This will allow you to match color in a RGB images for example.

Updated the documentation for filter2 to reflect this and added examples.

aoles commented on R/filter2.R:
```diff
@@ -45,10 +45,21 @@ filter2 = function(x, filter, boundary = c("circular", "replicate")) {
-           xpad = array(val, dx)
-           ## is there a better way of doing this?
-           x = do.call("[<-", c(quote(xpad), lapply(d, function(x) enquote(1:x)), quote(x)) )
+           x = imageData(x)
```
This call is obsolete, as the class is dropped by using the `array` to constructor

ghaarsma replied:
Agreed. Fixed in new commit to pull request.

aoles commented on R/filter2.R:
```diff
@@ -45,10 +45,21 @@ filter2 = function(x, filter, boundary = c("circular", "replicate")) {
-           x = do.call("[<-", c(quote(xpad), lapply(d, function(x) enquote(1:x)), quote(x)) )
+           x = imageData(x)
+           
+           dx[1:2] = dx[1:2] + df[1:2] - 1L
```
In the special case when we extend the canvas by filling it uniformly with a given value, there is no need to extend it from both sides. Please revert to `dx[1:2] = dx[1:2] + cf[1:2]` which creates a re-usable border shared in the left/right and top/bottom directions.

ghaarsma replied:
Please note that my pull request does not extend on both sides. It only pads on the right bottom. The do.call is functionally equivalent between our branches (we both pad on the right bottom side).

The code to extract the filtered image is identical as well:
```R
  if ( boundary!="circular" ) {
    y = asub(y, list(1:d[1L], 1:d[2L]), 1:2)
  }
```

The only difference is the amount of padding we do. In your case you pad a different amount between boundary=linear `(cf = df %/%2)` and boundary=replicate `(df-1)`. I think the padding size should be the same, only the padding values should be different, depending on the method.

The padding should be the size of the filter kernel minus the center point, therefore:
```R
dx[1:2] = dx[1:2] + df[1:2] - 1L
```
for both linear and replicate boundary options.

aoles replied:
Actually, my point was that the padding for `boundary = "linear"` doesn't need to be as wide as for the `"replicate"` mode. The reason is that in the former case the padding value is uniform around the image, therefore the same padding pixels can be reused when filtering at the left image edge, as when processing the right one (this is possible because the image is wrapped around the borders). You can verify that this indeed works by comparing the results from your code to my original implementation.

If you don't mind, I would prefer to keep the previous minimal padding, i.e. have
```r
dx[1:2] = dx[1:2] + cf[1:2]
```

aoles commented on R/filter2.R:
```diff
@@ -45,10 +45,21 @@ filter2 = function(x, filter, boundary = c("circular", "replicate")) {
+           
+           dx[1:2] = dx[1:2] + df[1:2] - 1L
+           
+           if ( length(dx)>2 & length(val)==prod(dx[-(1:2)]) ) {
```
Use `&&` instead of `&` in logical clauses

ghaarsma replied:
Agreed. Fixed in new commit to pull request.

aoles commented on R/filter2.R:
```diff
@@ -45,10 +45,21 @@ filter2 = function(x, filter, boundary = c("circular", "replicate")) {
+             xpad = array(val[1], dx)
+           }
+           # The do.call and arguments as a list of dimension is to take are of arrays of unknown dimension
+           x = do.call("[<-", c(list(xpad),lapply(d,FUN = function(x) 1:x),list(x)) )
```
What's the advantage of not quoting the arguments? My original implementation was protecting the arguments from early evaluation, which seems useful e.g. for debugging.

ghaarsma replied:
I took your comment to hart: ## is there a better way of doing this?  I didn't think you needed the quotes and were just costing overhead. 

I have looked at http://adv-r.had.co.nz/Computing-on-the-language.html for some inspiration, but I cannot think of a use case where quotes or enquote would be needed. It also debugs just fine.

All but the last argument cannot be formulas or expressions (they are always arrays/matrices or vectors). We could perhaps quote the last argument `x`. But not sure how you would use that.

aoles replied:
I don't think there is any measurable overhead by using the quotes, at least I was not able to confirm this. But maybe there is not much benefit of using the quoted expressions either. The advantage for debugging I had in mind was that `c(quote(xpad), lapply(d, function(x) enquote(1:x)), quote(x))` returns an unevaluated list of arguments, which I personally find more convenient to interact with.

aoles commented:
Hi Gabe,
thank you for your patience and for addressing my comments. I think we're almost good to go, please have a look at the remaining issue with the padding size.

Cheers,
Andrzej

ghaarsma commented:
Andrzej,

Updated pull request to address your two outstanding items.

1. Changed it back to quoted expressions in the do.call.  I was hoping that my implementation was slightly faster, but I checked with microbenchmark and the difference is insignificant.

2. Changed back to 'single' style padding for the linear boundary as you had it originally. Not quite sure what happened but I thought that your implementation was incorrect after I ran some tests. After trying again, I could not reproduce that and it turns out that your implementation should indeed just work fine.

Thanks for your patience,
Gabe

aoles commented:
Awesome, many thanks Gabe for contributing to the package. Your work is very much appreciated.

Cheers,
Andrzej
